### PR TITLE
Added support for trusted proxies / hosts in Request.

### PR DESCRIPTION
**Pull request by SamMousa**
- [x] Merge #13837 
- [x] Add test for `getUserIP`
- [x] Add test for `getMethod`
- [x] Add test for `getIsAjax`
- [x] Add test for `getIsPjax`

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #10854

### CODE REVIEW

**Comment by SamMousa** [comment]
Need some feedback on:
1. `yii\filters\HttpCache` uses `$_SERVER` directly. It currently has a dependency on `\Yii::$app->response`, can I just add a dependency on `\Yii::$app->request` to remove the usage of the global array?
2. Same for `ErrorHandler`.

**Comment by SamMousa** [comment]
Also, what is the convention on using getters vs magic properties inside the core?

In `Response` I see: `->getHeaders()`.

File: framework/web/Request.php
Comment (rob006) [inline framework/web/Request.php]: We could use simplified pattern, like in https://github.com/yiisoft/yii2/blob/35998449eba8e295f5ed420350765c1196bf13e3/framework/filters/HostControl.php#L68-L91

File: framework/web/Request.php
Comment (rob006) [inline framework/web/Request.php]: `is_numeric()` can return `true` for some intranet hosts. I suggest use `is_array($trustedHeaders)`?

Comment (rob006) [inline framework/web/Request.php]: Support for Closures could be also useful.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Hmm, so then the pattern is matched via `fnmatch`. That's possible as well, I'm just really against inventing new pattern types. 
I don't really see the benefit though. Every decent developer knows how to use regular expressions and they are more powerful.
The argument for `fnmatch` is more relevant when end users would be inputting patterns.

Comment (rob006) [inline framework/web/Request.php]: > That's possible as well, I'm just really against inventing new pattern types

We already using `fnmatch` for matching hosts and ips in some places, using regexp here is inconsistent. 

> Every decent developer knows how to use regular expressions and they are more powerful.

And they are tricky and less readable. Even in your example there is not escaped dot. :)

Comment (SamMousa) [inline framework/web/Request.php]: > And they are tricky and less readable. Even in your example there is not escaped dot. :)
Point taken :)

Still not a fan of `fnmatch` though.

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: What's the use case for trust only specific headers?

**Comment by samdark** [comment]
> yii\filters\HttpCache uses $_SERVER directly. It currently has a dependency on \Yii::$app->response, can I just add a dependency on \Yii::$app->request to remove the usage of the global array?

Yes.

> Same for ErrorHandler.

Yes.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Most proxies will ignore headers that they do not know about. So a proxy that doesn't set `X-Forwarded-Proto` but does set `X-Forwarded-For` will pass the other one through if the client sends it. This is insecure.

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: Not sure it makes sense to use separate set of rules for IP only...

File: framework/web/ErrorHandler.php
Comment (samdark) [inline framework/web/ErrorHandler.php]: One possible implication is about fatal errors with not enough memory... these could fail in the case. Not sure about it though. Needs testing.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: This is not a separate set of rules, but a configuration for `getUserIP`.
So step 1 is trusting certain headers. 
Step 2 is interpreting those headers (ie in which custom header is the users' IP stored?)

File: framework/web/ErrorHandler.php
Comment (SamMousa) [inline framework/web/ErrorHandler.php]: Yes that's why I was wondering if it was okay here to duplicate some of that logic.
On the other hand, the Request object gets created very early, so does the header collection. So they are most likely already in memory.
--> Not sure how to test this.

File: framework/filters/HttpCache.php
Comment (sergeymakinen) [inline framework/filters/HttpCache.php]: It looks like it should be

```php
if (Yii::$app->request->headers->has('If-None-Match'))
```

and so on.

Comment (SamMousa) [inline framework/filters/HttpCache.php]: @sergeymakinen Nice! will change that.

File: framework/web/ErrorHandler.php
Comment (samdark) [inline framework/web/ErrorHandler.php]: Cause out of memory error and see if it's logged.

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: This is not equivalent. Will break things.

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: This is not equivalent. Will break things.

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: Breaks BC because of method signature change.

Comment (samdark) [inline framework/web/Request.php]: I'd introduce separate method for new case.

File: tests/framework/web/RequestTest.php
Comment (samdark) [inline tests/framework/web/RequestTest.php]: Why it's left here?

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: This wasn't a separate method before. What would it be called and would it do more than store the result in a private variable?

Comment (samdark) [inline framework/web/Request.php]: Well, `checkHosts`? You've already named it in the comment here...

Comment (SamMousa) [inline framework/web/Request.php]: Actually that comment needs to be removed. all that stuff is in the `filterHeaders` function...

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: If there's no header, `null` will be returned and HHVM would produce "Warning: strcasecmp() expects parameter 1 to be string, null given" error. Either there should be `->has()` check or a default value as an empty string.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: That's right, that's why it is in the default configuration for `secureHeaders`. Meaning it will be used only if it the host is trusted.

Comment (samdark) [inline framework/web/Request.php]: That should be indicated clearly in phpdoc...

Comment (SamMousa) [inline framework/web/Request.php]: It is:
````
    /**
     * @var array The configuration for trusted security related headers.
     * An array key is a regular expression for matching a hostname.
     * A value is a list of headers to trust.
     *
     * For example, to trust all headers from domains ending in '.trusted.com' use the following:
     *
     * ```php
     * [
     *     '/^.*\.trusted\.com$/',
     * ```
     *
     * To trust just the x-forwarded-for header from domains ending in '.partial.com' use:
     *
     * ```
     *     // ...
     *     '/^.*\.partial\.com$/' => ['X-Forwarded-For']
     * ]
     * ```

     * Default is to trusts all headers from all hosts.
     */
````

Comment (SamMousa) [inline framework/web/Request.php]: Note that @rob006 made good arguments in favor of using `fnmatch` instead of `preg_match` for matching host names / ips.
What's your opinion?

File: framework/web/ErrorHandler.php
Comment (SamMousa) [inline framework/web/ErrorHandler.php]: I've confirmed that it still logs properly, both as normal and AJAX request.

Comment (samdark) [inline framework/web/ErrorHandler.php]: Should be fine then. Thanks for verifying it.

**Comment by SamMousa** [comment]
@samdark I'm looking into the CSRF tests now and I'm having some issues.
This is the original code:
````
/**
     * test CSRF token validation by POST param
     */
    public function testCsrfTokenHeader()
    {
        $this->mockWebApplication();
        $request = new Request();
        $request->enableCsrfCookie = false;
        $token = $request->getCsrfToken();
        // accept no value on GET request
        foreach (['GET', 'HEAD', 'OPTIONS'] as $method) {
            $_POST[$request->methodParam] = $method;
            $this->assertTrue($request->validateCsrfToken());
        }
        // only accept valid token on POST
        foreach (['POST', 'PUT', 'DELETE'] as $method) {
            $_POST[$request->methodParam] = $method;
            $request->setBodyParams([]);
            //$request->headers->remove(Request::CSRF_HEADER);
            unset($_SERVER['HTTP_' . str_replace('-', '_', strtoupper(Request::CSRF_HEADER))]);
            $this->assertFalse($request->validateCsrfToken());
            //$request->headers->add(Request::CSRF_HEADER, $token);
            $_SERVER['HTTP_' . str_replace('-', '_', strtoupper(Request::CSRF_HEADER))] = $token;
            $this->assertTrue($request->validateCsrfToken());
        }
    }
````

However with `enableCsrfCookie` set to `false`, the CSRF value will be saved in the session.
In my test setup, the session component does not work in phpunit due to output already started...
I've created a new test in `SessionTest`
````
public function testOpen() {
        $session = new Session();
        $session->open();
        $this->assertFalse($session->getIsActive());

        ini_set("session.use_cookies", 0);
        ini_set("session.use_only_cookies", 0);
        ini_set("session.cache_limiter", "");
        $session->open();
        $this->assertTrue($session->getIsActive());
}
````

This illustrates that session does not work in PHPUnit. 
It also indicates the tests for `Request` actually test `Session` as well.
One solution would be to mock a `Session` component that doesn't actually use session but instead just uses an array. This will work fine for unit tests.

Alternatively, `Request::getCsrfToken` could be used instead of `loadCsrfToken`, since we already assume an attacker cannot guess the CSRF token generated in a previous request, it is not unreasonable to assume an attack also cannot guess the CSRF token generated in the current request.

Please let me know how you'd like to proceed!

**Comment by samdark** [comment]
I'd use getter.

**Comment by SamMousa** [comment]
I've refactored the masking of CSRF tokens (related to the BREACH) vulnerability class.
- Use `Security::randomBytes`, even if such strong randomness is not needed, it does not hurt and makes the code easier to read.
- Introduced separate masking and unmasking functions that take care of masking and url encoding. Note that the unmasked token is never url encoded.
- Have added tests for the masking.

**Comment by samdark** [comment]
OK but that looks like totally unrelated. Would be great if it could be separated.

**Comment by SamMousa** [comment]
Yes I'll make a separate PR for that, it caused this PR to fail though, due to failing tests.

**Comment by SamMousa** [comment]
#13837

**Comment by SamMousa** [comment]
I've refactored the changes, will merge master back into this after #13837 is merged.

--> This might need to be squashed before merging 8-)

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: `@since 2.0.13`

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: `@since 2.0.13`

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: `2.0.13` or do you prefer `2.0.12`? ;)

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: `@since 2.0.13`

**Comment by SamMousa** [comment]
This is ready!

File: framework/web/Request.php
Comment (samdark) [inline framework/web/Request.php]: I don't think it's a good idea to make these secure by default. While it's probably fine to determine HTTP/HTTPS as in current master code, getting IP from these by default is not secure at all.

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: You misunderstand the meaning.
This is a list of headers that will NOT be added to the header collection unless it came from a trusted host.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Either is fine, i'll change to 2.0.13

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Note that the default configuration does mean that all hosts are trusted by default. It might make sense to instead just leave the default configuration empty (thus not trusting anyone).
Since in that it could be considered a security bug that it ever used untrusted headers for security related information.

It will mean that people who trusted the `https` headers blindly and used them for redirecting will need to reconfigure.

Comment (samdark) [inline framework/web/Request.php]: > This is a list of headers that will NOT be added to the header collection unless it came from a trusted host.

Then it should be named `$insecureHeaders`.

Comment (samdark) [inline framework/web/Request.php]: Or `$trustedOnlyHeaders`

Comment (samdark) [inline framework/web/Request.php]: Yes, even if things will break, it could be a good thing to do because it's about security...

**Review by samdark — APPROVED** [review]

**Review by dynasource — APPROVED** [review]
looks good to me. Well done.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: `get` -> `has`

Comment (cebe) [inline framework/web/Request.php]: may be faster in general to call get and check for null. Also in other places where isset and `$_SERVER` have been replaced.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: from above:

> Default is to trust all headers from all hosts.

you change the default value for userIP to be taken form a header by default, which was using the server remote address before, so installations that upgrade Yii will now become insecure by default.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: `@since` does not belong here, `getUserIP` existed before.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: this adds an unnecessary overhead of making DNS query to every request. In most cases this is not needed so it should only be performed if hostname is really required.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: the default should be to trust no headers and it should be explicitly configured to trust headers if application is installed behind a proxy.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: this would mean all headers are filtered that are not in trusted headers? That does not make sense. Headers should always be available, but trusted headers should be filtered if sent by ips that are not trusted.

**Review by cebe — CHANGES_REQUESTED** [review]

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: @cebe I agree, but this would severely break BC.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: This just gets the host from the header, it does not do DNS.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: That depends on the default value; with this config, which is the default:
````
    public $trustedHostConfig = [
        '//'
    ];
````

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: On second inspection, @cebe I think the comment is out of date. The default configuration will remove secure headers and allow all other headers from all hosts.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: It was requested here by @samdark a few comments up ^^

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Only if the host is trusted.
The comment in the source code regarding the default value is out of date. Will fix that.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: I think clear code is worth more. This path is only done once per request so the performance difference will be negligible.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: okay, I see now, but its not getting it from a header in your implementation, its using the `$_SERVER` variable provided by the webserver. We have an inconsistency in behavior then:

- `getUserHost()` will only use the hostname provided by the webserver
- `getUserIp()` will use headers or fallback to `getRemoteIp()`

I would expect `getUserHost()` to return the host of the IP returned by `getUserIp()`, which is not the case if that one is taken from a header.

A corresponding `getRemoteHost()` would be preferable.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: okay, this is fine then.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Hmm actually it should not use the user host at all. (Also that function doesn't exist).
It should be named `getRemoteHost()` (we're checking the next hop not the end user, the end user is never trusted to send us secure headers :p)

--> This means I need to double check the tests though; they should not have been passing.

**Comment by SamMousa** [comment]
@cebe I've updated the tests as well and have updated the `getRemoteHost()` documentation to reflect it's actual functionality.

So now you have:
`getUserIP()` --> get the users' IP.
`getRemoteIP()` --> get the next hop IP.
`getRemoteHost()` --> get the next hop host.

We do not have
`getUserHost()` --> it requires a DNS lookup.

Note that by default:
Apache does not provide `REMOTE_HOST` (https://httpd.apache.org/docs/2.4/mod/core.html#hostnamelookups)
Nginx doesn't provide it at all, as far as I can tell.

**Comment by SamMousa** [comment]
What happened to automated tests?
FYI: I've run the tests in `RequestTest` locally but not the whole suit.

**Comment by samdark** [comment]
Tests seems to be skipped if branch isn't up to date. Merged it w/ master and it seems to fix it.

**Comment by samdark** [comment]
Tests seems to pass.

**Comment by cebe** [comment]
For the purpose of introducing this without BC break and without introducing security issues:

Current behavior of Request class is the following:

- userIP: taken from `$_SERVER['REMOTE_ADDR']` currently, no header involved.
- userHost: taken from `$_SERVER['REMOTE_HOST']` currently, no header involved.
- isSecureConnection: Yii currently trusts the `X-Forwarded-Proto` header.

The default config in current state of this PR, will automatically trust `X-Forwarded-For` to retrieve the user IP. This should not be the default as it will make applications that rely on IP address for access checks insecure after updating Yii.

Also the `filterHeaders` implementation currently allows returning an empty list of headers when a host does not match any `trustedHostConfig` entry. I see no use case for this as it would result in very weird behavior of the application, which uses e.g. `Host:`, `X-Requested-With:` for ajax detection, `User-Agent`, etc. I do not think there is a reason to block these, at least on in terms of a "trusted proxy feature".

I'm going to propose some changes in a PR.

**Comment by SamMousa** [comment]
@cebe Feel free to directly commit to my PR branch; I think you have write permissions there automatically.

**Comment by cebe** [comment]
> @cebe Feel free to directly commit to my PR branch; I think you have write permissions there automatically.

done. @SamMousa @yiisoft/core-developers please review the current state.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: This will cause unexpected behavior when both host and ip are matched. Might not be an issue, but should at least be documented.

(Not something you introduced; just something I noticed just now)

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Should this not be a secure header as well?
Many security tools / proxies validate the URL using white / blacklists. (For example only allow `/admin` from the intranet)
This will allow an attack to bypass it.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: This naming is inconsistent with `$ipHeaders`.
Possibly rename `$ipHeaders` to `$secureIpHeaders` for consistency?

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Again not something introduced by you, but I found this in the source code for the PECL HTTP extension (where `http_get_request_headers()` comes from:

https://github.com/m6w6/ext-http/blob/45fb8dff8bc2981151fbca023408c61acbba0f08/src/php_http_env.c#L36

To me this suggests that anywhere this function is available `$_SERVER` must be available as well. Therefor I think this workaround could be removed.

**Review by SamMousa — COMMENTED** [review]
Minor stuff; thanks for helping me out. Looks a lot cleaner.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: "secure" in this case comes from `isSecureConnection`, but I'm not finally happy with the naming.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: we should probably break on first match.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: interesting find, but this belongs to a different issue.

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: good point, I have no idea how IIS works tough, this would be another BC break for systems that use this header.

Comment (SamMousa) [inline framework/web/Request.php]: Do we have any statistics on user base that uses IIS?

Comment (cebe) [inline framework/web/Request.php]: no

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: And document that we do so, that makes sense.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: Note that any functionality could simply be implemented without this header by simply using Yii routing rules.

All the webserver needs to do is get the request to `index.php`. So even removing the support altogether could be defended.

Comment (SamMousa) [inline framework/web/Request.php]: It's hard to find any documentation, but I think `X-Rewrite-Url` is not even part of IIS, but of a 3rd party plugin: http://www.helicontech.com/isapi_rewrite/

https://forums.asp.net/post/5807236.aspx

This could be resolved in a separate issue as well. I'll make one.

**Comment by SamMousa** [comment]
Is anymore needed for this PR?

**Comment by samdark** [comment]
@SamMousa these two questions:

- https://github.com/yiisoft/yii2/pull/13780/files#r125862700
- https://github.com/yiisoft/yii2/pull/13780/files#r125259774

Also the separate issues are to be created since when this one will be closed comments will be forgotten soon.

**Comment by SamMousa** [comment]
@samdark Have added this to my todo list; it might take me a while to get to it though; I'm swamped at the moment :(

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: This no long er applies. By default secure headers, which include `X-Forwarded-For` are not trusted and therefore the user IP is not taken from them unless trust has been configured.

File: framework/web/Request.php
Comment (SamMousa) [inline framework/web/Request.php]: I think this one is of out date; the current behavior is intuitive and logical.

**Comment by samdark** [comment]
What about breaking on 1st match in `filterHeaders()`?

**Comment by SamMousa** [comment]
@samdark I think I've handled all feedback, or replied in the comments. (@cebe requested changes to the docs but marked htem in the source code so github disagrees)

**Comment by samdark** [comment]
Merged. Thank you very much for it and for being patient addressing our comments.

File: framework/web/Request.php
Comment (krukru) [inline framework/web/Request.php]: Shouldn't this be break?

File: framework/web/Request.php
Comment (krukru) [inline framework/web/Request.php]: Wouldn't it be better to allow for subnet matching for the IP, rather than using regex. Or maybe even allow both. If the user provided a valid regular expression pattern, perform preg_match, otherwise check with respect to subnet (if the subnet is provided).

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: yes, good catch! Fixed: 9e713db

File: framework/web/Request.php
Comment (cebe) [inline framework/web/Request.php]: Discussion about different matching method has been discussed in https://github.com/yiisoft/yii2/pull/13780#discussion_r106126378 above already. I think its worth discussing that again.
Lets have a new issue for that.

Comment (cebe) [inline framework/web/Request.php]: -> https://github.com/yiisoft/yii2/issues/14691

**Pull request merged**
